### ISSUE TITLE
Issue with NFT

### ISSUE THREAD

**Issue by meme404erc** [Pandora-Labs-Org/erc404#4]
Hi,

we've deployed this new version of the contract but seems like there might be an issue with NFTs? Some users buying the token using Uniswap but do not see any NFT.

Etherscan link: https://etherscan.io/address/0x009bfab2a411d32954a180ae1c1e61d2cc43a4ac
Contract address: 0x009bfab2a411d32954a180ae1c1e61d2cc43a4ac

**Comment by myst4** [comment]
I have the same problem, mint some nft to test account, erc721 balance looks rigth but I cant see the nft on opensea for example

do you have any luck @meme404erc ?

**Comment by myst4** [comment]
adding more data, when I used erc404 old library (took from pandora example repo) it works greats on opensea when I minted some nft, but when I use this one erc404 library, it is missing on opensea after mint, anyone knows what is happens?

thanks

**Comment by meme404erc** [comment]
@myst4 unfortunately no, they are ignoring this question here and in builders Telegram chat :(

**Comment by mathdroid** [comment]
checking @meme404erc @myst4 , super swamped few days back

**Comment by mathdroid** [comment]
![image](https://github.com/Pandora-Labs-Org/erc404/assets/3748658/41bc9206-2395-4dc3-9949-d16a7c8446c1)
![image](https://github.com/Pandora-Labs-Org/erc404/assets/3748658/960099d1-2609-4428-8e95-e5181ec432e0)
![image](https://github.com/Pandora-Labs-Org/erc404/assets/3748658/3ad8c8c4-8274-4a11-a13d-754e815061b6)

I am not seeing any problems onchain. Maybe the issue is in Opensea?

**Comment by mathdroid** [comment]
![image](https://github.com/Pandora-Labs-Org/erc404/assets/3748658/5567e871-55d4-40e1-a7a5-4a593f4e3d21)
![image](https://github.com/Pandora-Labs-Org/erc404/assets/3748658/1af99f36-dd7c-4e7f-bfb8-6533cece579b)
![image](https://github.com/Pandora-Labs-Org/erc404/assets/3748658/50462d3c-03ca-4350-81a8-fa6303125d7d)
![image](https://github.com/Pandora-Labs-Org/erc404/assets/3748658/fab4085f-8574-4840-83a3-dcfe67b2534e)
![image](https://github.com/Pandora-Labs-Org/erc404/assets/3748658/dc8601c9-fbac-4bec-b6a0-ce8b63a58fe2)
![image](https://github.com/Pandora-Labs-Org/erc404/assets/3748658/9e1ed343-9087-4cf4-8e89-9e897adb7db1)
![image](https://github.com/Pandora-Labs-Org/erc404/assets/3748658/39b9cb8d-bb0a-4e2f-a5b3-276f25a723fe)

so they HAVE the tokens, but there was no erc721transfer event in the last pic. I think this used an old version of the code. cc @caldereth @0xacme

**Comment by mathdroid** [comment]
need more e2e test: transfer from exempted pool @caldereth

**Comment by myst4** [comment]
@mathdroid  yeap the nft looks fine in the contract, but in opensea or any nft market cant read the contract, 
I tried with the older erc404 library and after transfer or swap it appers in real time in opensea, but with this new erc404 library it not works on opensea

this is the nft contract builded with erc404 old (the same that is using pandora token)

```
https://opensea.io/assets/matic/0xc3a608429487b6dbb1177ec930f77b71e53cb4d4/3
```


and I followd the same steps using the example404.sol from this repo and I seems that not mints the nfts, just transfer all as erc20 tokens for that reason seems that opensea cant read it

```
https://opensea.io/assets/matic/0xb94647FE572A9f26Fe60c9BcEe1bE24320A9878f/2
```

it throws 404 error on opensea

any one idea ?

**Comment by myst4** [comment]
> ![image](https://private-user-images.githubusercontent.com/3748658/304179110-5567e871-55d4-40e1-a7a5-4a593f4e3d21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NjMxNjQsIm5iZiI6MTcwNzc2Mjg2NCwicGF0aCI6Ii8zNzQ4NjU4LzMwNDE3OTExMC01NTY3ZTg3MS01NWQ0LTQwZTEtYTdhNS00YTU5M2Y0ZTNkMjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTJUMTgzNDI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWJiNzdlNDA3NjIzN2M4ZjhmNzljY2ZkMTZhMjhkOWQxNTI0NmVmOGFjOTE1OGY1OWZmOGQ4Mjg2MzUxMDU2ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.pqiFTTeDG6v8D9eyCM4WGGIyznRCw97EhOJtoSYySN4) ![image](https://private-user-images.githubusercontent.com/3748658/304179197-1af99f36-dd7c-4e7f-bfb8-6533cece579b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NjMxNjQsIm5iZiI6MTcwNzc2Mjg2NCwicGF0aCI6Ii8zNzQ4NjU4LzMwNDE3OTE5Ny0xYWY5OWYzNi1kZDdjLTRlN2YtYmZiOC02NTMzY2VjZTU3OWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTJUMTgzNDI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGQwNGRlYTlkYjU1MjBlOGRlZmNiNzIxNDI0NDE2MjE1ZjAyMWNmMjNkNDJlYTJlM2RjN2FiNmZjN2JlZTU4ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ApT9pTdc1ltGvOc6-p4Sd0AgYDjdJtWuqOE_HVFQLbE) ![image](https://private-user-images.githubusercontent.com/3748658/304179495-50462d3c-03ca-4350-81a8-fa6303125d7d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NjMxNjQsIm5iZiI6MTcwNzc2Mjg2NCwicGF0aCI6Ii8zNzQ4NjU4LzMwNDE3OTQ5NS01MDQ2MmQzYy0wM2NhLTQzNTAtODFhOC1mYTYzMDMxMjVkN2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTJUMTgzNDI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDE2MTJhYWM1ZmIxMTkyZDU2NzkxODJkZWY2OTM3MGIwY2VmMDcyODI2NTBjMjUxYmQ3ZGI0NjIzNTUyNjhhYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.5MQCEaRMAopKVAMmRYQH3-KItaiD30A1Zou3_0_XAmk) ![image](https://private-user-images.githubusercontent.com/3748658/304179556-fab4085f-8574-4840-83a3-dcfe67b2534e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NjMxNjQsIm5iZiI6MTcwNzc2Mjg2NCwicGF0aCI6Ii8zNzQ4NjU4LzMwNDE3OTU1Ni1mYWI0MDg1Zi04NTc0LTQ4NDAtODNhMy1kY2ZlNjdiMjUzNGUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTJUMTgzNDI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTc5NjQ2NmMxMDA0Y2I0ZDBiOTEyZDY0M2JmNjAzNGFmYTZmM2Y1YjBiMzU0MzU3M2YwZGFiYTYzMGNiYTczNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ceioCFW4AWFxCVbDPf93WoVGrDtP9SzCFJUnK6Btz3Q) ![image](https://private-user-images.githubusercontent.com/3748658/304179642-dc8601c9-fbac-4bec-b6a0-ce8b63a58fe2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NjMxNjQsIm5iZiI6MTcwNzc2Mjg2NCwicGF0aCI6Ii8zNzQ4NjU4LzMwNDE3OTY0Mi1kYzg2MDFjOS1mYmFjLTRiZWMtYjZhMC1jZThiNjNhNThmZTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTJUMTgzNDI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODhlMmViM2UwZjA0MDkzZjNmODNlNmQ5OGEwMDM5MWVlZTM0ZmI1YzRmZDkzOTFiY2RhNzlhZTM0NzlmYjU5MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.txjGk_iSJFq4t3GJyWHKrYtOIWDtrbG90GjTgLRr9ZQ) ![image](https://private-user-images.githubusercontent.com/3748658/304179739-9e1ed343-9087-4cf4-8e89-9e897adb7db1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NjMxNjQsIm5iZiI6MTcwNzc2Mjg2NCwicGF0aCI6Ii8zNzQ4NjU4LzMwNDE3OTczOS05ZTFlZDM0My05MDg3LTRjZjQtOGU4OS05ZTg5N2FkYjdkYjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTJUMTgzNDI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWMzNzdiNDZjMDVhMzhhNGU1YjA2NDcwMTI4ZWY1YTdmNjdjNzUyM2Q0NWU4NjFkODk1OWFiNTliZTJkZTEzYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.OY94Tfo5YUTjDomN84O_Dmc0ndZqSRpohO0HmvHxwh4) ![image](https://private-user-images.githubusercontent.com/3748658/304180062-39b9cb8d-bb0a-4e2f-a5b3-276f25a723fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NjMxNjQsIm5iZiI6MTcwNzc2Mjg2NCwicGF0aCI6Ii8zNzQ4NjU4LzMwNDE4MDA2Mi0zOWI5Y2I4ZC1iYjBhLTRlMmYtYTViMy0yNzZmMjVhNzIzZmUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTJUMTgzNDI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzBlZmVlYTQ4ZWM5ZjQ1M2I1YmNjZGQ3YWFjZDczZGQzZTljOTYwMDk3NzFhNTAyNjNlYTFmNGE4ODYwNDlmZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.8JvdrShtkMRTobRaeJi7BsIZHpvvx6nRdls4tlbkXGM)
> 
> so they HAVE the tokens, but there was no erc721transfer event in the last pic. I think this used an old version of the code. cc @caldereth @0xacme

exactly, on the contracts looks all ok but never mint the nft

check this transfer 

```
https://etherscan.io/tx/0xf7ff8c3c19381e5f5581dd59cb74ce2150f8df4fe5d60864cee533ce007abfc7
```

the user send 1667000000000000000 tokens (1,6) so it should mint 1 nft, but just sent all as erc20 tokens.

**Comment by mathdroid** [comment]
![image](https://github.com/Pandora-Labs-Org/erc404/assets/3748658/4def033e-908f-4e40-b9af-f40ba98098d7)
in #6 we have a test for this and it passed (events were emitted). will add a more thorough case @myst4 

cc @0xacme @caldereth I think we should use old audited 404 in `main` branch, develop the new one in `canary` or other branch

**Comment by celsosa** [comment]
Same here, no nfts are being minted (and contract shows like the wallet have the nfts). We need to add a custom mint nft fuction on transfer or something? Or should work already in current protocol code?

**Comment by mathdroid** [comment]
@celsosa use v1 for now while we fix this

**Comment by myst4** [comment]
@mathdroid  where is the last stable version of erc404?, is  in some branch in this repo?

**Comment by gpsus** [comment]
Does anyone have a solution to this?

**Comment by smarter777** [comment]
@myst4 , @meme404erc 
I faced once this issue and solved on my end.
Main reason of NFT is not be shown on Opensea is due to `Transfer` event.
In ERC404-legacy version, there's event "Transfer" and emitted whenever ERC721 token is transferred, but not in ERC404 v2.
Instead there's ERC721Transfer event.
By adding Transfer event, I solved this issue because Opensea monitor the "Transfer" event, not ERC721Transfer event.
Depending on your mind whether you would use Opensea or the others for NFT trading, but I sincerely recommend to this solution if you want to use Opensea NFT marketplace.
Hope this solution would be help for you.
